File: 2024/01/day01.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
    "math"
)


func calculations(filename string) {
    // Read file
    file, err := os.Open(filename)
    if err != nil {
        log.Fatal(err)
    }
    defer file.Close()

    // Init result
    var result1 int = 0
    var result2 int = 0
    locationID1 := []int {}
    locationID2 := []int {}

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        line := scanner.Text()
        // Code goes here
        parts := strings.Split(line, "   ")

        num1, err := strconv.Atoi(parts[0])
        if err != nil {
            log.Fatal(err)
        }
        locationID1 = append(locationID1, num1)

        num2, err := strconv.Atoi(parts[1])
        if err != nil {
            log.Fatal(err)
        }
        locationID2 = append(locationID2, num2)
    }

    fmt.Println("Part One ####################")
    sort.Slice(locationID1, func(i, j int) bool {
        return locationID1[i] < locationID1[j]
    })
    sort.Slice(locationID2, func(i, j int) bool {
        return locationID2[i] < locationID2[j]
    })

    for i := range locationID1 {
        result1 += int(math.Abs(float64(locationID1[i] - locationID2[i])))
    }

    fmt.Printf("Result: %d\n", result1)

    fmt.Println("Part Two ####################")
    freqs := make(map[int]int)
    for _, v := range locationID2 {
        freqs[v]++
    }

    for _, v := range locationID1 {
        mult, exists := freqs[v]
        if !exists {
            mult = 0
        }
        result2 += v*mult
    }

    fmt.Printf("Result: %d\n", result2)
}

func main() {
    if len(os.Args[1:]) < 1 {
        log.Fatal("Not enough arguments given!")
    }

    filename := os.Args[1]
    calculations(filename)
}

File: 2024/01/day01.py
#!/usr/bin/python3
import sys
from typing import List


def part_one(lines: List[str]):
    print('Part One')
    locs = [[], []]
    for l in lines:
        a, b = l.split("   ")
        locs[0].append(int(a))
        locs[1].append(int(b))

    for l in locs:
        l.sort()

    r1 = 0
    for a, b in zip(*locs):
        r1 += abs(a - b)
    print(r1)

    print('Part Two')

    r2 = 0
    freqs = dict()
    for e in locs[1]:
        if e not in freqs:
            freqs[e] = 0
        freqs[e] += 1

    for v in locs[0]:
        r2 += v*freqs.get(v, 0)
    print(r2)

def main(args):
    if len(args) < 2:
        print('[ERROR] Enter a file name!')
        print(f'\tUsage: python {args[0]} <filename>')
        sys.exit(1)

    filename = args[1]
    with open(filename, 'r') as f:
        lines = f.readlines()
        lines = [l.strip() for l in lines]

    part_one(lines)


if __name__ == '__main__':
    main(sys.argv)
